File: app/src/androidTest/java/it/sephiroth/android/library/uigestures/demo/TestTapGesture.java
package it.sephiroth.android.library.uigestures.demo;

import android.graphics.Point;
import android.os.SystemClock;
import android.support.test.espresso.action.ViewActions;
import android.support.test.espresso.matcher.ViewMatchers;
import android.support.test.filters.SdkSuppress;
import android.support.test.runner.AndroidJUnit4;
import android.support.test.uiautomator.UiObject;
import android.support.test.uiautomator.UiObjectNotFoundException;

import junit.framework.Assert;

import org.junit.Test;
import org.junit.runner.RunWith;

import it.sephiroth.android.library.uigestures.UIGestureRecognizer.State;
import it.sephiroth.android.library.uigestures.UIGestureRecognizerDelegate;
import it.sephiroth.android.library.uigestures.UITapGestureRecognizer;

import static android.support.test.espresso.Espresso.onView;
import static org.junit.Assert.assertEquals;

@RunWith (AndroidJUnit4.class)
@SdkSuppress (minSdkVersion = 18)
public class TestTapGesture extends TestBaseClass {

    @Test
    public void test_singleTap() throws UiObjectNotFoundException, InterruptedException {
        BaseTest activity = activityTestRule.getActivity();

        final UIGestureRecognizerDelegate delegate = activity.delegate;
        Assert.assertNotNull(delegate);
        delegate.clear();

        UITapGestureRecognizer recognizer = new UITapGestureRecognizer(context);
        recognizer.setTag("single-tap");
        recognizer.setNumberOfTouchesRequired(1);
        recognizer.setNumberOfTapsRequired(1);
        recognizer.setActionListener(activityTestRule.getActivity());
        delegate.addGestureRecognizer(recognizer);

        getTitleView().setText("1 Tap");
        getTextView().setText("None");

        onView(ViewMatchers.withId(R.id.activity_main)).perform(ViewActions.click());
        SystemClock.sleep(2000);

        assertEquals(recognizer.getTag() + ": " + State.Ended, getTextView().getText());
    }

    @Test
    public void test_singleTap2Fingers() throws UiObjectNotFoundException, InterruptedException {
        final UIGestureRecognizerDelegate delegate = activityTestRule.getActivity().delegate;
        Assert.assertNotNull(delegate);
        delegate.clear();

        UITapGestureRecognizer recognizer = new UITapGestureRecognizer(context);
        recognizer.setTag("single-tap");
        recognizer.setNumberOfTouchesRequired(2);
        recognizer.setNumberOfTapsRequired(1);
        recognizer.setActionListener(activityTestRule.getActivity());
        delegate.addGestureRecognizer(recognizer);

        final UiObject mainView = getMainView();

        getTitleView().setText("1 Tap 2 Fingers");

        mainView.performTwoPointerGesture(
            new Point(200, 300),
            new Point(200, 400),
            new Point(200, 300),
            new Point(200, 400),
            1
        );
        SystemClock.sleep(200);
        assertEquals(recognizer.getTag() + ": " + State.Ended, getTextView().getText());
    }

    @Test
    public void test_singleTap2Taps() throws UiObjectNotFoundException, InterruptedException {
        BaseTest activity = activityTestRule.getActivity();

        final UIGestureRecognizerDelegate delegate = activity.delegate;
        delegate.clear();

        UITapGestureRecognizer recognizer = new UITapGestureRecognizer(context);
        recognizer.setTag("single-tap");
        recognizer.setNumberOfTouchesRequired(1);
        recognizer.setNumberOfTapsRequired(2);
        recognizer.setActionListener(activityTestRule.getActivity());
        delegate.addGestureRecognizer(recognizer);

        getTitleView().setText("2 Taps");
        getTextView().setText("None");

        onView(ViewMatchers.withId(R.id.activity_main)).perform(ViewActions.doubleClick());
        SystemClock.sleep(200);

        assertEquals(recognizer.getTag() + ": " + State.Ended, getTextView().getText());
    }

}

File: app/src/androidTest/java/it/sephiroth/android/library/uigestures/demo/TestTapGesture.kt
package it.sephiroth.android.library.uigestures.demo

import android.graphics.Point
import androidx.test.espresso.Espresso.onView
import androidx.test.espresso.action.ViewActions
import androidx.test.espresso.matcher.ViewMatchers
import it.sephiroth.android.library.uigestures.UIGestureRecognizer.State
import it.sephiroth.android.library.uigestures.UITapGestureRecognizer
import junit.framework.Assert
import org.junit.Assert.assertEquals
import org.junit.Test
import org.junit.runner.RunWith
import java.util.concurrent.CountDownLatch

@RunWith(androidx.test.ext.junit.runners.AndroidJUnit4::class)
class TestTapGesture : TestBaseClass() {

    @Test
    fun test_singleTap() {
        val latch = CountDownLatch(1)
        val activity = activityTestRule.activity
        val delegate = activity.delegate
        Assert.assertNotNull(delegate)
        delegate.clear()

        val recognizer = UITapGestureRecognizer(context)
        recognizer.tag = "single-tap"
        recognizer.touchesRequired = 1
        recognizer.tapsRequired = 1
        recognizer.actionListener = {
            assertEquals(State.Ended, it.state)
            latch.countDown()
        }
        delegate.addGestureRecognizer(recognizer)

        textView.text = "None"

        onView(ViewMatchers.withId(R.id.activity_main)).perform(ViewActions.click())
        latch.await()
    }

//    @Test
    fun test_singleTap2Fingers() {
        val latch = CountDownLatch(1)
        val delegate = activityTestRule.activity.delegate
        Assert.assertNotNull(delegate)
        delegate.clear()

        val recognizer = UITapGestureRecognizer(context)
        recognizer.tag = "single-tap"
        recognizer.touchesRequired = 2
        recognizer.tapsRequired = 1
        recognizer.actionListener = activityTestRule.activity.actionListener

        recognizer.actionListener = {
            assertEquals(State.Ended, it.state)
            latch.countDown()
        }

        delegate.addGestureRecognizer(recognizer)

        mainView.performTwoPointerGesture(
                Point(100, 200),
                Point(200, 300),
                Point(100, 200),
                Point(200, 300),
                1)

        latch.await()
    }

    @Test
    fun test_doubleTap() {
        val latch = CountDownLatch(1)
        val activity = activityTestRule.activity

        val delegate = activity.delegate
        delegate.clear()

        val recognizer = UITapGestureRecognizer(context)
        recognizer.tag = "double-tap"
        recognizer.touchesRequired = 1
        recognizer.tapsRequired = 2
        recognizer.tapTimeout = 400
        recognizer.actionListener = {
            assertEquals(State.Ended, it.state)
            latch.countDown()
        }

        delegate.addGestureRecognizer(recognizer)

        textView.text = "None"

        onView(ViewMatchers.withId(R.id.activity_main)).perform(ViewActions.doubleClick())
        latch.await()
    }

}
